sphinx_doc/tutorials/docs/writing_new_costmap2d_plugin.rst

```diff
@@ -0,0 +1,195 @@
+
+In current implementation there is also setting ``need_recalculation_`` bounds recalculation indicator:
+
+.. code-block:: c
+
+   need_recalculation_ = false;
+
+2. ``GradientLayer::updateCosts()`` method is called by the ``LayeredCostmap`` each time when costmap re-calculation is required. It updates the costmap layer only within its bounds. There are 4 input parameters of method: calculation window bounds and 1 output parameter: reference to ``GradientLayer`` costmap. Inside this method it is placed the gradient making algorithm for underlying costmap:
+
+.. code-block:: c
+
+  int gradient_index;
+  for (int j = min_j; j < max_j; j++) {
+    // Reset gradient_index each time when reaching the end of re-calculated window
+    // by OY axis.
+    gradient_index = 0;
+    for (int i = min_i; i < max_i; i++) {
+      int index = master_grid.getIndex(i, j);
+      // assign the cost associated with the distance from an obstacle to the cell
+      unsigned char cost = (LETHAL_OBSTACLE - gradient_index*GRADIENT_FACTOR)%255;
+      if (gradient_index <= GRADIENT_SIZE) {
+        gradient_index++;
+      } else {
+        gradient_index = 0;
+      }
+      master_array[index] = cost;
+    }
+  }
+
+where the ``GRADIENT_SIZE`` is the size of each gradient period in map cells, ``GRADIENT_FACTOR`` - decrement of costmap's value per each step:
+
+.. image:: images/Writing_new_Costmap2D_plugin/gradient_explanation.png
+
+These parameters are defined in plugin's header file.
+
+3. ``GradientLayer::updateBounds()`` is called by the ``LayeredCostmap`` to ask the plugin: which area of costmap layer it needs to update. There are 3 input parameters of method: robot position and orientation and 4 output parameters: pointers to window bounds. In current implementation ``updateBounds()`` re-calculates window bounds if ``need_recalculation_`` is ``true`` and updates them independently on ``need_recalculation_`` value.
+
+4. ``GradientLayer::onFootprintChanged()`` method is called whenever the footprint was changed. In current implementation it just resets ``need_recalculation_`` value.
+
+Last ``reset()`` method is dummy:
+
+5. ``GradientLayer::reset()`` is not used in this example plugin. It remaining there since pure virtual function ``reset()`` in parent ``Layer`` class required to be overriden. However in another plugins implementations it may have any code to be executed during costmap reset.
+
+2- Enable the plugin in Costmap2D
+---------------------------------
+
+1. At first, plugin's class should be linked with parent in ``costmap_plugins.xml``:
+
+.. code-block:: diff
+
+  @@ -12,6 +12,9 @@
+       <class type="nav2_costmap_2d::VoxelLayer"     base_class_type="nav2_costmap_2d::Layer">
+         <description>Similar to obstacle costmap, but uses 3D voxel grid to store data.</description>
+       </class>
+  +    <class type="nav2_costmap_2d::GradientLayer"  base_class_type="nav2_costmap_2d::Layer">
```

Comment:
Since this will be a standalone package, this should talk about what the costmap_plugins.xml file is and then adding this into it (and what each of the arguments mean)